### PR TITLE
[6.x] Implement Pet Specializations

### PR DESCRIPTION
**Pull request by myran2**
**Changes proposed**:
- Removed pet talent code, this isn't used in 6.x
- Implemented pet spec. related packets
- Added an extra column to character_pet to store the pet's specialization

**Target branch(es)**: 6x

**Issues addressed**: 

**Tests performed**: (Does it build, tested in-game, etc): Builds

**Known issues and TODO list**:
- [x] The pet action bar isn't instantly updated after a spec change. A button on the bar has to be clicked first.
- [x] After choosing a pet specialization, the UI looks like this http://puu.sh/oDLZs.jpg however, after relogging, the UI reverts back to this http://puu.sh/oDM48.jpg although the pet still has its specialization spells.There's probably some flag missing that has to be sent on login.

### CODE REVIEW

File: sql/updates/characters/6.x/2016_MM_DD_00_characters.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `character_pet` ADD COLUMN `specialization` SMALLINT UNSIGNED NOT NULL DEFAULT '0' AFTER `abdata`;
```
Comment (Aokromes) [inline sql/updates/characters/6.x/2016_MM_DD_00_characters.sql]: missing newline.

File: src/server/game/Spells/Auras/SpellAuraEffects.cpp [hunk 2]
```diff
@@ -2619,7 +2619,8 @@ void AuraEffect::HandleAuraUntrackable(AuraApplication const* aurApp, uint8 mode
 
 void AuraEffect::HandleAuraModPetTalentsPoints(AuraApplication const* aurApp, uint8 mode, bool /*apply*/) const
```
Comment (tkrokli) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp]: What causes this fatal error in Travis?
```
src/server/game/Spells/Auras/SpellAuraEffects.cpp:2620:71: fatal error: 
      unused parameter 'aurApp' [-Wunused-parameter]
void AuraEffect::HandleAuraModPetTalentsPoints(AuraApplication const* aurApp...
                                                                      ^
```
Is it caused by a change in this PR or something outside of it?

Comment (myran2) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp]: The contents of the function are commented out so all of the arguments are unused. That's what's throwing the error.

Comment (tkrokli) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp]: Ah, right. I should have noticed. Thanks for pointing it out.

**Comment by Lordron** [comment]
https://github.com/TrinityCore/TrinityCore/pull/17058/commits/787a730f0fc86d44e87e3e1fdc13b7fc32eece60#diff-046ce69d5dd9f7bc5d2212637e38a233R1792

@myran2 maybe it will be better to unlearn spells only from current spec (if present) rather than going through all specs

**Comment by Shauren** [comment]
@Lordron no, this is good and prevents weird behavior in case spec id is set to 0 before removing spells

File: sql/updates/characters/6.x/2016_MM_DD_00_characters.sql (hunk 1 above)
Comment (tkrokli) [inline sql/updates/characters/6.x/2016_MM_DD_00_characters.sql]: Added in https://github.com/TrinityCore/TrinityCore/pull/17058/commits/d43d54e99750b9fb1af27e8bca26957f49e94441 .

**Comment by tkrokli** [comment]
Looks like you just need to `insert the missing names of the already applied sql updates to the 'updates' table of the 'characters' base dump ('sql/base/characters_database.sql')` to make this PR build successfully in Travis. Does this PR need anything else before it is ready?

**Comment by myran2** [comment]
This still hasn't been tested in game. That definitely needs to be done before merging.

Also, I'd like to add support for the AT_LOGIN_RESET_PET_TALENTS flag if it is still used (the name should probably be changed to reflect that it changes pet spec. instead of talents though) and remove the pet talents aura effect after I confirm that it is no longer used.

**Comment by Shauren** [comment]
Aura effect 145 (previously SPELL_AURA_MOD_PET_TALENT_POINTS) is used by 5 spells in 6.2.4 dbc but the meaning of this aura changed (it's used by mind control spells but isn't the control itself)

**Comment by sarnex** [comment]
Just all around garbage honestly

**Comment by myran2** [comment]
It seems like stabling a pet deletes it so it's kind of hard to test the third TODO entry. Is that a consequence of this PR or is that the current behavior on master? I'll fix this issue either way, but if it's not caused by this PR I'd like to open a separate one to fix  the issue.

File: src/server/game/Entities/Pet/Pet.cpp [hunk 3]
```diff
@@ -1940,3 +1795,71 @@ void Pet::ResetGroupUpdateFlag()
...
+}
+
+void Pet::RemoveSpecializationSpells(bool clearActionBar)
+{
+    std::vector<uint32> unlearnedSpells;
+
+    for (uint32 i = 0; i < MAX_PET_SPECIALIZATIONS; ++i)
+    {
+        // pets have class id 0 in ChrSpecialization.dbc
+        if (ChrSpecializationEntry const* specialization = sChrSpecializationByIndexStore[PET_SPEC_OVERRIDE_CLASS_INDEX][i])
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.cpp]: i meant clearing both sChrSpecializationByIndexStore[PET_SPEC_OVERRIDE_CLASS_INDEX][i] and sChrSpecializationByIndexStore[0][i] here (2x this if block)

File: src/server/game/Entities/Pet/Pet.cpp [hunk 4]
```diff
@@ -1940,3 +1795,71 @@ void Pet::ResetGroupUpdateFlag()
...
+    }
+
+    learnSpells(learnedSpells);
+}
+
+void Pet::RemoveSpecializationSpells(bool clearActionBar)
+{
+    std::vector<uint32> unlearnedSpells;
+
+    for (uint32 i = 0; i < MAX_PET_SPECIALIZATIONS; ++i)
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.cpp]: just use MAX_SPECIALIZATIONS, there is no need for additional define (especially that you iterate array with MAX_SPECIALIZATIONS size (4))

File: sql/updates/characters/6.x/2016_MM_DD_00_characters.sql [hunk 5]
```diff
@@ -0,0 +1 @@
+--ALTER TABLE `character_pet` ADD COLUMN `specialization` SMALLINT UNSIGNED NOT NULL DEFAULT '0' AFTER `abdata`;
```
Comment (Shauren) [inline sql/updates/characters/6.x/2016_MM_DD_00_characters.sql]: perhaps this shouldn't be commented out

File: src/server/database/Database/Implementation/CharacterDatabase.cpp [hunk 6]
```diff
@@ -649,17 +649,19 @@ void CharacterDatabaseConnection::DoPrepareStatements()
...
     PrepareStatement(CHAR_UPD_CHAR_PET_SLOT_BY_SLOT_EXCLUDE_ID, "UPDATE character_pet SET slot = ? WHERE owner = ? AND slot = ? AND id <> ?", CONNECTION_ASYNC);
     PrepareStatement(CHAR_UPD_CHAR_PET_SLOT_BY_SLOT, "UPDATE character_pet SET slot = ? WHERE owner = ? AND slot = ?", CONNECTION_ASYNC);
     PrepareStatement(CHAR_UPD_CHAR_PET_SLOT_BY_ID, "UPDATE character_pet SET slot = ? WHERE owner = ? AND id = ?", CONNECTION_ASYNC);
     PrepareStatement(CHAR_DEL_CHAR_PET_BY_ID, "DELETE FROM character_pet WHERE id = ?", CONNECTION_ASYNC);
-    PrepareStatement(CHAR_DEL_CHAR_PET_BY_SLOT, "DELETE FROM character_pet WHERE owner = ? AND (slot = ? OR slot > ?)", CONNECTION_ASYNC);
+    PrepareStatement(CHAR_DEL_CHAR_PET_BY_OWNER, "DELETE FROM character_pet WHERE owner = ? AND (slot = ? OR slot > ?)", CONNECTION_ASYNC);
+    PrepareStatement(CHAR_DEL_ALL_PET_SPELLS_BY_OWNER, "DELETE FROM `pet_spell` WHERE `guid` in (SELECT `id` FROM `character_pet` WHERE owner=?)", CONNECTION_ASYNC);
+    PrepareStatement(CHAR_UPD_PET_SPECS_BY_OWNER, "UPDATE character_pet SET specialization = 0 WHERE owner=?)", CONNECTION_ASYNC);
```
Comment (Shauren) [inline src/server/database/Database/Implementation/CharacterDatabase.cpp]: typo in the query

File: src/server/game/Entities/Pet/Pet.cpp [hunk 7]
```diff
@@ -1436,6 +1432,25 @@ bool Pet::learnSpell(uint32 spell_id)
...
+        if (!addSpell(spell))
+            continue;
+
+        packet.Spells.push_back(spell);
+    }
+
+    if (!m_loading)
+    {
+        GetOwner()->GetSession()->SendPacket(packet.Write());
+        //GetOwner()->PetSpellInitialize();
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.cpp]: don't add new code that is commented out

File: src/server/game/Entities/Pet/Pet.cpp [hunk 8]
```diff
@@ -1940,3 +1795,82 @@ void Pet::ResetGroupUpdateFlag()
...
+            SpecializationSpellsEntry const* specSpell = specSpells->at(j);
+            SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(specSpell->SpellID);
+            if (!spellInfo || spellInfo->SpellLevel > getLevel())
+                continue;
+
+            learnedSpells.push_back(specSpell->SpellID);
+        }
+    }
+
+    learnSpells(learnedSpells);
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.cpp]: `learnSpells(std::move(learnedSpells));`

File: src/server/game/Entities/Pet/Pet.cpp [hunk 9]
```diff
@@ -1940,3 +1795,82 @@ void Pet::ResetGroupUpdateFlag()
...
+                for (size_t j = 0; j < specSpells->size(); ++j)
+                {
+                    SpecializationSpellsEntry const* specSpell = specSpells->at(j);
+                    unlearnedSpells.push_back(specSpell->SpellID);
+                }
+            }
+        }
+    }
+
+    unlearnSpells(unlearnedSpells, true, clearActionBar);
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.cpp]: `unlearnSpells(std::move(unlearnedSpells), true, clearActionBar);`

File: src/server/game/Entities/Pet/Pet.cpp [hunk 10]
```diff
@@ -446,7 +442,7 @@ void Pet::SavePetToDB(PetSaveMode mode)
 
         // save pet
         std::ostringstream ss;
-        ss  << "INSERT INTO character_pet (id, entry,  owner, modelid, level, exp, Reactstate, slot, name, renamed, curhealth, curmana, abdata, savetime, CreatedBySpell, PetType) "
+        ss  << "INSERT INTO character_pet (id, entry,  owner, modelid, level, exp, Reactstate, slot, name, renamed, curhealth, curmana, abdata, savetime, CreatedBySpell, PetType, specialization) "
```
Comment (joschiwald) [inline src/server/game/Entities/Pet/Pet.cpp]: can we have a prepared statement here?

File: src/server/game/Entities/Pet/Pet.cpp (hunk 9 above)
Comment (ProjectcoreDevs) [inline src/server/game/Entities/Pet/Pet.cpp]: Missing PET_SPEC_OVERRIDE_CLASS_INDEX definition :)

Comment (myran2) [inline src/server/game/Entities/Pet/Pet.cpp]: https://github.com/TrinityCore/TrinityCore/commit/7a806595ee916bef13783d12e0fbee4df0ff804b

File: src/server/game/Handlers/PetHandler.cpp [hunk 11]
```diff
@@ -734,35 +734,28 @@ void WorldSession::HandlePetSetSpecializationOpcode(WorldPackets::Pet::LearnPetS
...
-        default:
-            TC_LOG_ERROR("network", "WorldSession::HandlePetSetSpecializationOpcode: Invalid SpecGroupIndex (%d) supplied.", packet.SpecGroupIndex);
-            return;
+    if (packet.SpecGroupIndex >= MAX_SPECIALIZATIONS)
+    {
+        TC_LOG_DEBUG("network", "WORLD: HandlePetSetSpecializationOpcode - specialization index %u out of range", packet.SpecGroupIndex);
+        return;
+    }
+
+    int specIndex = _player->HasAuraType(SPELL_AURA_OVERRIDE_PET_SPECS) ? PET_SPEC_OVERRIDE_CLASS_INDEX : 0;
```
Comment (Shauren) [inline src/server/game/Handlers/PetHandler.cpp]: use uint32

File: src/server/game/Entities/Pet/Pet.h [hunk 12]
```diff
@@ -167,6 +169,7 @@ class TC_GAME_API Pet : public Guardian
 
         DeclinedName *m_declinedname;
 
+        uint8 m_specGroup;
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.h]: You don't need to store it - you can find it in ChrSpecializationEntry::OrderIndex (and you can get ChrSpecializationEntry by a simple dbc lookup with m_petSpecialization)

File: src/server/game/Spells/Auras/SpellAuraEffects.cpp [hunk 13]
```diff
@@ -6651,3 +6651,32 @@ void AuraEffect::HandleModSpellCategoryCooldown(AuraApplication const* aurApp, u
...
+        ChrSpecializationEntry const* petSpec = sChrSpecializationByIndexStore[PET_SPEC_OVERRIDE_CLASS_INDEX][pet->GetSpecGroup()];
+        pet->SetSpecialization(petSpec->ID);
+    }
+    else
+    {
+        ChrSpecializationEntry const* petSpec = sChrSpecializationByIndexStore[0][pet->GetSpecGroup()];
+        pet->SetSpecialization(petSpec->ID);
+    }
+
+}
```
Comment (Shauren) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp]: missing newline at EOF

File: src/server/game/Entities/Pet/Pet.cpp [hunk 14]
```diff
@@ -1940,3 +1789,90 @@ void Pet::ResetGroupUpdateFlag()
...
+            }
+        }
+    }
+
+    unlearnSpells(unlearnedSpells, true, clearActionBar);
+}
+
+void Pet::SetSpecialization(uint16 spec)
+{    
+    if (m_petSpecialization == spec);
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.cpp]:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Pet/Pet.cpp:1849:37: fatal error: 
      if statement has empty body [-Wempty-body]
    if (m_petSpecialization == spec);
```
notice the semicolon

File: src/server/game/Entities/Pet/Pet.cpp (hunk 9 above)
Comment (sarnex) [inline src/server/game/Entities/Pet/Pet.cpp]: Amateur hour, a new feature film starring @ProjectcoreDevs and directed by fuck you

File: src/server/game/Entities/Pet/Pet.h [hunk 15]
```diff
@@ -38,6 +38,16 @@ enum PetStableinfo
     PET_STABLE_INACTIVE = 2
 };
 
+enum PetTalents
```
Comment (Shauren) [inline src/server/game/Entities/Pet/Pet.h]: This enum isn't needed anymore

**Comment by Aokromes** [comment]
Any reason to don't merge this? Travis thing is updates thing, needs to be added at merge.

**Comment by myran2** [comment]
As far as I know, everything is working right now. There could be an issue with non-active pets retaining the adapted specialization IDs after the hunter loses the Adaptation talent but I haven't had time to test it.

**Pull request closed without merging**